test/e2e/gateway_api_test.go

```diff
@@ -423,6 +428,286 @@ func testGatewayAPIRBAC(t *testing.T) {
+		if err := kclient.Delete(context.TODO(), &gateway); err != nil {
+			if errors.IsNotFound(err) {
+				return
+			}
+			t.Errorf("failed to delete gateway %q: %v", gateway.Name, err)
+		}
+	})
+
+	if err != nil {
+		t.Fatalf("failed to create gateway with a listener of hostname foo.%s: %v", domain, err)
+	}
+
+	gateway, err = assertGatewaySuccessful(t, operatorcontroller.DefaultOperandNamespace, "test-gateway-update")
+	if err != nil {
+		t.Fatalf("failed to create the test gateway: %v", err)
+	}
+
+	t.Logf("Created gateway %s with hostname %s", gateway.Name, "foo."+domain)
+
+	if err := assertExpectedDNSRecords(t, map[expectedDnsRecord]bool{
+		{dnsName: "foo." + domain + ".", gatewayName: "test-gateway-update"}: true,
+	}); err != nil {
+		t.Fatalf("DNSRecord expectations not met: %v", err)
+	}
+
+	// Fetch the latest version of the Gateway resource.
+	if err := kclient.Get(context.TODO(), types.NamespacedName{Namespace: operatorcontroller.DefaultOperandNamespace, Name: gateway.Name}, gateway); err != nil {
+		t.Fatalf("failed to get gateway resource: %v", err)
+	}
+
+	// Modify one gateway listener hostname
+	newHostname := gatewayapiv1.Hostname("baz." + domain)
+	gateway.Spec.Listeners[0].Hostname = &newHostname
+
+	if err := kclient.Update(context.TODO(), gateway); err != nil {
+		t.Fatalf("failed to update gateway %v: %v", gateway.Name, err)
+	}
+	t.Logf("Modified gateway's %s listener hostname to: %s", gateway.Name, "baz."+domain+".")
+
+	if err := assertExpectedDNSRecords(t, map[expectedDnsRecord]bool{
+		{dnsName: "baz." + domain + ".", gatewayName: "test-gateway-update"}: true,
+	}); err != nil {
+		t.Fatalf("DNSRecord expectations not met: %v", err)
+	}
+
+	// Fetch the latest version of the Gateway resource.
+	if err := kclient.Get(context.TODO(), types.NamespacedName{Namespace: operatorcontroller.DefaultOperandNamespace, Name: gateway.Name}, gateway); err != nil {
+		t.Fatalf("failed to get gateway resource: %v", err)
+	}
+
+	// Delete one of the listeners
+	var updatedListeners []gatewayapiv1.Listener
+	for _, l := range gateway.Spec.Listeners {
+		if l.Name == "http-listener2" {
+			continue
+		}
+		updatedListeners = append(updatedListeners, l)
+	}
+	gateway.Spec.Listeners = updatedListeners
+	if err := kclient.Update(context.TODO(), gateway); err != nil {
+		t.Fatalf("failed to update gateway with removed listener: %v", err)
+	}
+	t.Logf("Deleted listener %s from gateway %s", "http-listener2", gateway.Name)
+
+	t.Logf("Checking that the dnsRecord for %s gets removed.", "baz."+domain+".")
+
+	if err := assertExpectedDNSRecords(t, map[expectedDnsRecord]bool{
+		{dnsName: "baz." + domain + ".", gatewayName: "test-gateway-update"}: true,
+		{dnsName: "foo." + domain + ".", gatewayName: "test-gateway-update"}: false,
+		{dnsName: "bar." + domain + ".", gatewayName: "test-gateway-update"}: false,
```

Comment:
nit:
Maybe it would be better to check for each after making the changes, change the `assertExpectedDNSRecord` to return the bool.  Not a deal-breaker.